src/main.ml

```diff
@@ -71,6 +71,7 @@ let scan_workspace ?(log=Log.no_log)
   }
 
 let init_build_system ?only_packages ?external_lib_deps_mode w =
+  Context.set w.contexts;
```

Comment:
Don't we run this multiple times in a polling build?

Reply:
Ah yes. We should have some better automated tests for them. I don't use them myself. In any case, this can easily be fixed, so I'd like the approach to be validated first.

Reply:
I've fixed this.

Reply:
Sorry, I should have said more words.

If something is global and mutable, then any memoized computation that depends on it can return an incorrect result, something this PR is trying to avoid in the first place.

I guess a solution could be to make `Context.db` a "memo-variable", a memoized function that we'll invalidate every time we "set" its value. This is similar to how we think about filesystem access. (except somebody else does the "set"ting)

Currently we reset all caches every run anyway, so it's not going to change behavior.
Maybe we should leave a comment though, saying what form we intend this module to take in the future.

Reply:
> I guess a solution could be to make Context.db a "memo-variable", a memoized function that we'll invalidate every time we "set" its value. This is similar to how we think about filesystem access. (except somebody else does the "set"ting)

Sounds good to me. I would also expect that a memoized function that depends on  the context would take it explicitly as an argument, but that could be hard to enforce.

Reply:
> I would also expect that a memoized function that depends on the context would take it explicitly as an argument

Waait a minute, isn't that the opposite of what you're doing in this PR (you're removing the argument)?

Reply:
I don't expect those functions to be memoized. And if we'd like to memoize a part of them, I'd extract that part and make sure Context.t, profile, etc. are all passed to the memoized functions.

Reply:
Well, they are currently already memoized using lazy, it seems, and that's what lead you to make this change. Memoization using lazy is not compatible with the other memoization so we will almost certainly have to change it sooner or later. If you're saying passing context explicitly is where you'd want to end up then, then I don't understand why make this PR which moves you in the opposite direction.

Reply:
Lazy? Perhaps you mean the explicit mutability. My main motivation was to get rid of the profile argument as it makes the API brittle.

Did you mean to get rid of mutability in Env_node in favor of memorization? In that case that would make sense. I agree with your original point then.

Reply:
Btw, the memorization I had in mind was for example doing it on the function that constructs the flags. In my mind, it would take the context, the set of flags in the parent dir, and the current env config for example. I'm not entirely sure how things are going to look like here however

Reply:
> Lazy? Perhaps you mean the explicit mutability. 

I didn't read the code, so I don't know, but there was at least one "lazy" keyword around.

> Did you mean to get rid of mutability in Env_node in favor of memorization?

Yeah, basically.

> I'm not entirely sure how things are going to look like here however

Fair enough. We could change nothing for now then? I'm not sure this PR accomplishes its goal. (if anything, it might make things harder to understand in the future because the dependency on global variable is implicit)